include/util/static_rtree.hpp

```diff
@@ -166,15 +167,245 @@ class StaticRTree
     StaticRTree &operator=(const StaticRTree &) = delete;
 
-    // Construct a packed Hilbert-R-Tree with Kamel-Faloutsos algorithm [1]
+    enum PackingMethod
+    {
+        Hilbert,
+        STR
+    };
+
     explicit StaticRTree(const std::vector<EdgeDataT> &input_data_vector,
                          const std::string &tree_node_filename,
                          const std::string &leaf_node_filename,
-                         const Vector<Coordinate> &coordinate_list)
+                         const Vector<Coordinate> &coordinate_list,
+                         const PackingMethod packing_method = Hilbert)
         : m_coordinate_list(coordinate_list)
     {
+        util::Log() << "Starting rtree";
+        switch (packing_method)
+        {
+        case Hilbert:
+            PackWithHilbert(input_data_vector, tree_node_filename, leaf_node_filename);
+            break;
+        case STR:
+            PackWithSTR(input_data_vector, tree_node_filename, leaf_node_filename);
+            break;
+        default:
+            PackWithHilbert(input_data_vector, tree_node_filename, leaf_node_filename);
+        }
+    }
+
+    // Construct a packed STR-RTree with the Luetengger-Edgington-Lopez approach
+    // STR performs better than Hilbert for road-network-like data where the
+    // distribution is only mildly biased
+    void PackWithSTR(const std::vector<EdgeDataT> &input_data_vector,
+                     const std::string &tree_node_filename,
+                     const std::string &leaf_node_filename)
+    {
+        util::Log() << "Packing with STR";
+        auto copy = const_cast<std::vector<EdgeDataT> &>(input_data_vector);
+        //        copy.reserve(input_data_vector.size());
+        // std::copy(input_data_vector.begin(), input_data_vector.end(), copy);
+
+        // Sort by longitude coordinate of centroid for each segment
+        // TODO: this means calculating centroids a lot, this could be moved into
+        // a separate single pass
+        util::Log() << "Sorting leaves by centroid longitude";
+        tbb::parallel_sort(
+            copy.begin(), copy.end(), [this](const EdgeDataT &a, const EdgeDataT &b) {
+                auto a_centroid = coordinate_calculation::centroid(m_coordinate_list[a.u],
+                                                                   m_coordinate_list[a.v]);
+                auto b_centroid = coordinate_calculation::centroid(m_coordinate_list[b.u],
+                                                                   m_coordinate_list[b.v]);
+                return a_centroid.lon < b_centroid.lon;
+
+            });
+
+        // Now, break into even-sized vertical slices, and sort each slice by latitude
+        // TODO: make sure this isn't 0
+        const std::size_t leaf_K = std::sqrt(copy.size() / LEAF_NODE_SIZE);
+        util::Log() << "leaf_K is " << leaf_K;
+        tbb::parallel_for(tbb::blocked_range<uint64_t>(0, copy.size(), leaf_K),
```

Comment:
I would expect [floor(sqrt(15809/127))] = 11 vertical stripes for monaco but the result is
![str](https://cloud.githubusercontent.com/assets/4421046/25178638/fa4e1182-2506-11e7-9230-71da7552e168.png)

The problem that `leaf_K` is used as `grainsize` by `tbb::blocked_range` and is not what expected.